### ISSUE TITLE
float32 data type

### ISSUE THREAD

**Issue by xiorcale** [nlpodyssey/spago#53]
This is just a question out of curiosity but: Do you have any plans to support the `float32` data type (or any other types, like integers actually)? 

- It is very common to train a neural network with a `float32` precision, as it reduces the computation cost without having any significant impact on the accuracy, and I was wondering what would be the speed gain for spago?
- I was thinking about something like an `Enum` type given to the matrix creation function and maybe the possibility to convert an existing matrix from a type to another
- Supporting types like `uint8` could allow to implement some quantization schemes more easily, which seems like a good fit with Golang since it shines in distributed applications (e.g. sending quantized weights over the network is definitely more bandwidth friendly) 

For now I just find myself "fighting the matrix" by extracting the underlying data, convert them in the desired type, do some work with them and finally reloading them later in a matrix to run some computation.

**Comment by matteo-grella** [comment]
Hi @xiorcale,

You have raised a very interesting point. When I started writing spaGO, I used `float64` without thinking about it too much. I thought it was a safe choice not to find myself with precision issues any later.

However, as you say, PyTorch and other ML frameworks use `float32` by default as the impact in neural models is negligible. In fact, when loading pre-trained Hugging Face models we force the conversion from `float32` to `float64`.

The question now is whether it makes sense to keep `float64` at all or just convert everything to `float32`. What do you think?
We could then measure the speed increase with a benchmark on a dedicated branch.

Where to enable support for integers is another important design decision. In general, I agree with you, especially on the quantization. Just recently I asked myself this question, implementing in spaGO the recent Google's PRADO model (https://ai.googleblog.com/2020/09/advancing-nlp-with-efficient-projection.html).

Let me know if you are interested in contributing to this topic!

**Comment by matteo-grella** [comment]
@jjviana do you have any opinion here?

**Comment by matteo-grella** [comment]
@marco-nicola

**Comment by xiorcale** [comment]
For me it would make sense to have a "typed matrix", which can be whatever the user requires (`float64`, `float32`, `int`, ...). Maybe some inspiration can be taken from https://github.com/gorgonia/tensor?

Regarding contribution, I'm not sure to find the time right now and I probably need to explore spago a bit more before trying to modify it, but if I find anything worth contributing, be sure I'll do it.

 As a side note, I saw that gonum has assembly code for `float32` which could probably be used here in the same way you did for `float64` (?)

**Comment by jjviana** [comment]
As long as spago is cpu-only, I don't see a lot of advantages of moving to float32 (except maybe memory space). 
x86 supports doubles natively and will execute floating point instructions in 64 bits with the same speed as 32 bits. 
If GPU support is required then it is a completely different story, but that would also mean a larger refactoring.

**Comment by marco-nicola** [comment]
Hey,
we have some news on this topic.

We tried to replace float64 with float32 on the whole project, and it turned out that many high-level tasks (e.g. BART and BERT classifications) execute in almost half of the time!

While single operations performance didn't probably change much, not to mention that we are still converting values to float64 for using functions from Go's `math` package, more data can fit in the processor's cache. At least that's my guess, I'm not really an expert on that level.

On top of that, we couldn't observe any significant loss of precision on the overall results. Apparently a win-win!

So we decided to switch to float32 as default data type on spaGO. Yet we didn't want to throw away float64 completely. At this point, making the two data types really interchangeable at runtime would easily end up in cumbersome code and structures. Generics might help us here, but we still have to wait for them.

Eventually, as a lest-painful solution, we implemented two packages: `mat32` and `mat64`. They provide an identical public API (functions and types, including matrices) for almost every mathematical operation used by the other packages. You can easily "switch" between 32 and 64 bits by simply importing one or the other package. By aliasing those packages as simply `mat` when they are imported, no further changes are required. We provided a very simple bash script which replaces the imports for you: `change-float-type.sh`.

These changes are included in the latest release `v0.3.0`.

We'd really appreciate your opinions and feedback!

Of course there's still some space for even better performances, from float32-native functions, to more intelligent use of SIMD instructions... but that's a different story.

Cheers!

**Comment by xiorcale** [comment]
Wow, this looks really promising, I'm glad to hear about these results 🎉 

I'm not really surprised about the accuracy staying the same, but I was really curious to see the impact regarding execution time, which seems to be significant enough to make your work worth it 😉 

Actually, I don't have much to add, as you pretty much summarized it: without generic, switching the type at runtime is probably a no-go. Your solution, even though not as flexible as what can be found in other programming language seems to be reasonable for Golang.

Regarding the conversion in `float64` in order to use the `math` package, I know that the creator of Gorgonia has a reimplementation in `float32`: https://github.com/chewxy/math32 (but I never took a deep look at it though...).

Again thanks for this amazing work, I will definitely try to play with the `float32` implementation and will come back here if I find any remark worth to mention.

**Comment by jjviana** [comment]
This looks really promising indeed!
This is one of the cases where theory and practice diverge - in theory,
there should be not a lot of improvement in speed by going from float64 to
float32 (given the underlying operations are 64bit anyway), but in
practice, the memory access pattern improvements make it worth it.

I agree that creating two separate packages is probably the best way of
doing that in Go.

I will give 0.3.0 a try with Golem and will let you know if I run  into
anything.

Regards,
 - Juliano
